Cookbook/Technical-Documents/AutomateAUIFlow.md

```diff
@@ -0,0 +1,259 @@
+
+It is at this level where we should be asserting screen states, grouping *cross screen* actions or completing a single action. A detailed naming convention for step definitions is defined [here](https://github.com/Babylonpartners/ios-playbook/blob/master/Cookbook/Technical-Documents/UIAutomation.md).
+
+### Screen Objects
+Screen objects are intended to contain all code that controls the user interface, checks the screen state or in limited cases groups *single screen* actions. Functions within a Screen objects should adhere to the single responsibility principle and should never reference another screen object.
+
+#### State Checks
+State check functions are intended to *evaluate* the screen under tests and return a value. In practise these fall, into three categories but are not limited to these. Code to check if something is displayed, normally returning a Boolean. 
+
+```swift 
+func isScreenDisplayed() -> Bool
+``` 
+
+We can with well written step definitions pass in a parameter to be evaluated. For example passing in the name of **Card** to verified as displayed on the *HomeScreen*.
+
+```swift
+func isCardDisplayed(for card: HomeScreenCards) -> Bool
+``` 
+
+The last check which we normally perform is the fetch a value from the screen, for example to fetch how many Pharamcies are being displayed on screen.
+
+```swift 
+func pharmaciesCount() -> Int
+```
+
+What ever is being returned should be evaluated at the Step Definition level and not within the screen object itself.
+
+```swift
+XCTAssertTrue(HomeScreen().isScreenDisplayed(), "Home screen icon was not displayed")
+```
+
+#### Interface Interactions
+Interactions are intended to cover user actions on the screen, if we look at the login screen we would find functions like 
+
+```swift 
+func enterEmail(_ email: String)
+func enterPassword()
+``` 
+
+We can also group common actions so long as they are within the scope of the screen and make logical sense, staying with the login screen we could have: 
+
+```swift 
+func enterCredentialsAndLogin(email: String)
+```
+
+#### Accessibility identifiers
+The historical method for handling Accessibility identifiers that we're going to be used in locating an **XCUIElement** was to define and stored within the screen object, normally in a *enum* for each element type. And then referenced from the functions using ```Cells.passwordField```. This ensured that identifiers are only stored once in test code and can be updated with a single change.
+
+This however is now evolved, with rather than having Accessibility identifiers define in production code and again in the Screen Object, we now access the identifier defined in a **AccessibilityContent** enum in the production code and reference from the automation code. ```AccessibilityContent.SignUp.termsAndConditionsCell``` this ensures identifiers are only defined and read from one location. Though is a work in progress.
```

Comment:
```suggestion
This however is now evolved, with rather than having Accessibility identifiers defined in production code and again in the Screen Object, we now access the identifier defined in a **AccessibilityContent** enum in the production code and reference from the automation code. ```AccessibilityContent.SignUp.termsAndConditionsCell``` this ensures identifiers are only defined and read from one location. Though is a work in progress.
```